src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -1927,31 +1915,26 @@ public TokenMetadata getTokenMetadata()
         return map;
     }
 
+    public InetAddressAndPort getNativeAddressAndPort(InetAddressAndPort endpoint)
+    {
+        NodeInfo<?> info = Nodes.localOrPeerInfo(endpoint);
+        if (info != null && info.getNativeTransportAddressAndPort() != null)
+            return info.getNativeTransportAddressAndPort();
+
+        if (endpoint.equals(FBUtilities.getBroadcastAddressAndPort()))
+            return FBUtilities.getBroadcastNativeAddressAndPort();
+
+        return InetAddressAndPort.getByAddressOverrideDefaults(endpoint.address, DatabaseDescriptor.getNativeTransportPort());
+    }
+
     /**
      * Return the native address associated with an endpoint as a string.
      * @param endpoint The endpoint to get rpc address for
      * @return the native address
      */
-    public String getNativeaddress(InetAddressAndPort endpoint, boolean withPort)
+    public String getNativeAddress(InetAddressAndPort endpoint, boolean withPort)
     {
-        if (endpoint.equals(FBUtilities.getBroadcastAddressAndPort()))
-            return FBUtilities.getBroadcastNativeAddressAndPort().getHostAddress(withPort);
-        else if (Gossiper.instance.getEndpointStateForEndpoint(endpoint).getApplicationState(ApplicationState.NATIVE_ADDRESS_AND_PORT) != null)
-        {
-            try
-            {
-                InetAddressAndPort address = InetAddressAndPort.getByName(Gossiper.instance.getEndpointStateForEndpoint(endpoint).getApplicationState(ApplicationState.NATIVE_ADDRESS_AND_PORT).value);
-                return address.getHostAddress(withPort);
-            }
-            catch (UnknownHostException e)
-            {
-                throw new RuntimeException(e);
-            }
-        }
-        else if (Gossiper.instance.getEndpointStateForEndpoint(endpoint).getApplicationState(ApplicationState.RPC_ADDRESS) == null)
```

Comment:
Are these RPC_ADDRESS else if's covered in the new logic?

Reply:
yes, see this method: https://github.com/datastax/cassandra/pull/300/files#diff-398cb35f0a0e84f5142f64c40ca2d0280803e4b2b39478409d2656ff8e594a69R310